cmd/tsconnect/wasm/wasm_js.go

```diff
@@ -324,6 +350,117 @@ func (i *jsIPN) logout() {
 	go i.lb.Logout()
 }
 
+func (i *jsIPN) tcp(host string, port int, readCallback js.Value, connectTimeoutSeconds int) any {
+	portInt := strconv.Itoa(port)
+	addr := net.JoinHostPort(host, portInt)
+
+	return makePromise(func() (any, error) {
+		if readCallback.Type() != js.TypeFunction {
+			return nil, errors.New("Invalid readCallback received")
+		}
+
+		connectTimeout := time.Duration(connectTimeoutSeconds) * time.Second
+		jsTCPSession, err := i.NewJSTCPSession(addr, readCallback, connectTimeout)
+		if err != nil {
+			return nil, err
+		}
+
+		jsObjectWrapper := jsTCPSession.jsWrapper()
+		return jsObjectWrapper, nil
+	})
+
+}
+
+type jsTCPSession struct {
+	jsIPN       *jsIPN
+	conn        net.Conn
+	writeBuffer []byte
+	readBuffer  []byte
+}
+
+const defaultTCPConnectTimeout time.Duration = 5 * time.Second
+
+func (i *jsIPN) NewJSTCPSession(addr string, readCallback js.Value, connectTimeout time.Duration) (*jsTCPSession, error) {
+	if connectTimeout == 0 {
+		connectTimeout = defaultTCPConnectTimeout
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), connectTimeout)
+	defer cancel()
+
+	conn, err := i.dialer.UserDial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &jsTCPSession{
+		jsIPN:       i,
+		conn:        conn,
+		writeBuffer: make([]byte, 1024),
+		readBuffer:  make([]byte, 4*1024*1024),
+	}
+
+	go s.readLoop(readCallback)
+
+	return s, nil
+}
+
+func (s *jsTCPSession) readLoop(readCallback js.Value) {
+	defer s.conn.Close()
+	dst := js.Global().Get("Uint8Array").New(len(s.readBuffer))
+
+	// we always reuse the same dst but make use of subarrays
+	subarray := dst.Call("subarray", 0, len(s.readBuffer))
+	subarrayLength := len(s.readBuffer)
+	for {
+		n, err := s.conn.Read(s.readBuffer[:])
+		if err != nil {
+			if err == io.EOF {
+				return
+			}
+			log.Printf("read error: %v", err)
+			return
+		}
+		// if subarray is not already the correct length,
+		// create a new one from dst with correct len
+		if subarrayLength != n {
+			subarrayLength = n
+			subarray = dst.Call("subarray", 0, n)
+		}
+
+		js.CopyBytesToJS(subarray, s.readBuffer[:n])
+
+		readCallback.Invoke(subarray)
+	}
+}
+
+func (s *jsTCPSession) jsWrapper() any {
+	return map[string]any{
+		"close": js.FuncOf(func(this js.Value, args []js.Value) interface{} {
```

Comment:
use "any" for all "interface{}".

Also, can you add little comments briefly describing the signatures of these JS funcs (whatever args is supposed to be). It can just be in TypeScript syntax if you prefer.